common/lib/plugins/efm2/monitor.ts

```diff
@@ -0,0 +1,311 @@
+  private readonly properties: Map<string, any>;
+  private readonly hostInfo: HostInfo;
+  private stopped: boolean = false;
+
+  private monitoringClient: ClientWrapper | null = null;
+
+  private readonly failureDetectionTimeNano: number;
+  private readonly failureDetectionIntervalNanos: number;
+  private readonly failureDetectionCount: number;
+
+  private invalidHostStartTimeNano: number;
+  private failureCount: number;
+  private hostUnhealthy: boolean = false;
+  private readonly abortedConnectionsCounter: TelemetryCounter;
+
+  constructor(
+    pluginService: PluginService,
+    hostInfo: HostInfo,
+    properties: Map<string, any>,
+    failureDetectionTimeMillis: number,
+    failureDetectionIntervalMillis: number,
+    failureDetectionCount: number,
+    abortedConnectionsCounter: TelemetryCounter
+  ) {
+    this.pluginService = pluginService;
+    this.telemetryFactory = this.pluginService.getTelemetryFactory();
+    this.hostInfo = hostInfo;
+    this.properties = properties;
+    this.failureDetectionTimeNano = failureDetectionTimeMillis * 1000000;
+    this.failureDetectionIntervalNanos = failureDetectionIntervalMillis * 1000000;
+    this.failureDetectionCount = failureDetectionCount;
+    this.abortedConnectionsCounter = abortedConnectionsCounter;
+
+    const hostId: string = this.hostInfo.hostId ?? this.hostInfo.host;
+
+    this.telemetryFactory.createGauge(`efm2.newContexts.size.${hostId}`, () => MonitorImpl.newContexts.size == Number.MAX_SAFE_INTEGER);
+
+    this.telemetryFactory.createGauge(`efm2.activeContexts.size.${hostId}`, () => this.activeContexts.length == Number.MAX_SAFE_INTEGER);
+
+    this.telemetryFactory.createGauge(`efm2.hostHealthy.${hostId}`, () => (this.hostUnhealthy ? 0 : 1));
+    const task1 = this.newContextRun();
+    const task2 = this.run();
+    Promise.race([task1, task2]).finally(() => {
+      this.stopped = true;
+    });
+  }
+
+  canDispose(): boolean {
+    return this.activeContexts.length === 0 && MonitorImpl.newContexts.size === 0;
+  }
+
+  startMonitoring(context: MonitorConnectionContext): void {
+    if (this.isStopped()) {
+      logger.warn(Messages.get("MonitorImpl.monitorIsStopped", this.hostInfo.host));
+    }
+
+    const currentTimeNanos: number = this.getCurrentTimeNano();
+    const startMonitorTimeNano = currentTimeNanos + this.failureDetectionTimeNano;
+    const connectionQueue = MapUtils.computeIfAbsent(
+      MonitorImpl.newContexts,
+      startMonitorTimeNano,
+      () => new Array<WeakRef<MonitorConnectionContext>>()
+    );
+    connectionQueue.push(new WeakRef<MonitorConnectionContext>(context));
+  }
+
+  async newContextRun(): Promise<void> {
+    logger.debug(Messages.get("MonitorImpl.startMonitoringTaskNewContext", this.hostInfo.host));
+
+    try {
+      while (!this.isStopped()) {
+        const currentTimeNanos = this.getCurrentTimeNano();
+        // Get entries with key (that is a time in nanos) less or equal than current time.
```

Comment:
Can we double check this? The actual check is `key < currentTimeNanos`, do we need to update this comment or the if statement?